src/sagemaker_containers/_intermediate_output.py

```diff
@@ -0,0 +1,165 @@
+
+        # delete the original file
+        os.remove(file_path)
+    except FileNotFoundError:
+        # Broken link or deleted
+        pass
+    except Exception:
+        logger.exception('Failed to upload file to s3.')
+
+        # delete the original file
+        os.remove(file_path)
+
+
+def _copy_file(executor, s3_uploader, relative_path, filename):
+    try:
+        src = os.path.join(intermediate_path, relative_path, filename)
+        dst = os.path.join(tmp_dir_path, relative_path, '{}.{}'.format(_timestamp(), filename))
+        shutil.copy2(src, dst)
+        executor.submit(_upload_to_s3, s3_uploader, relative_path, dst, filename)
+    except FileNotFoundError:
+        # Broken link or deleted
+        pass
+    except Exception:
+        logger.exception('Failed to copy file to the temporarily directory.')
+
+
+def _watch(inotify, watchers, watch_flags, s3_uploader):
+    # initialize a thread pool with 1 worker
+    # to be used for uploading files to s3 in a separate thread
+    executor = ThreadPoolExecutor(max_workers=1)
+
+    last_pass_done = False
+    stop_file_exists = False
+
+    # after we see stop file do one additional pass to make sure we didn't miss anything
+    while not last_pass_done:
+        # wait for any events in the directory for 1 sec and then re-check exit conditions
+        for event in inotify.read(timeout=1000):
+            for flag in flags.from_mask(event.mask):
+                # if new directory was created traverse the directory tree to recursively add all
+                # created folders to the watchers list.
+                # Upload files to s3 if there any files.
+                # There is a potential race condition if upload the file and the see a notification
+                # for it which should cause any problems because when we copy files to temp dir
+                # we add a unique timestamp up to microseconds.
+                if flag is flags.ISDIR:
+                    for folder, dirs, files in os.walk(os.path.join(intermediate_path, event.name)):
+                        wd = inotify.add_watch(folder, watch_flags)
+                        relative_path = os.path.relpath(folder, intermediate_path)
+                        watchers[wd] = relative_path
+                        os.makedirs(os.path.join(tmp_dir_path, relative_path))
+                        for file in files:
+                            _copy_file(executor, s3_uploader, relative_path, file)
+                elif flag is flags.CLOSE_WRITE:
+                    _copy_file(executor, s3_uploader, watchers[event.wd], event.name)
+
+        last_pass_done = stop_file_exists
+        stop_file_exists = os.path.exists(success_file_path) or os.path.exists(failure_file_path)
+
+    # wait for all the s3 upload tasks to finish and shutdown the executor
+    executor.shutdown(wait=True)
+
+
+def start_intermediate_folder_sync(s3_output_location, region):
```

Comment:
When you name a function in Python, always think about module + fn because nobody should import just your fn name.

For example, this one fullname is **_intermediate_output. start_intermediate_folder_sync**

Reply:
+1